### DISCUSSION TITLE
Support for Transformation Functions (FnO)

### DISCUSSION THREAD

**Discussion by nraphael** [morph-kgc/morph-kgc#69] [Ideas]
Since RML is itself an extension of R2RML, it is likely you have thought of this yourself. Or indeed, have already implemented it and it is my bad for not having seen it.

Anyway, if you haven't, it seems reasonable (to me :-) ) that, depending on the complexity of intermediate transformations between the inputs and the RDF outputs, to want to encode those extensions inline as a particular property of the various maps. A new property, perhaps, rr:transform, say, that would name a function that would get called with the following parameters:

- the current input field
- the whole record, indexed by input field
- the current map
- the entire map
- Others: constants section for each map, each logical source and the whole dataset ...?

Yes, one can add post-morph transformations in python, where then each mapping file is part of a mapping script -- but I prefer (and maybe this is just a personal preference) the declarative mode and would like to stay in it as long as possible. It is "self-documenting" to a degree.

Of course I could transform my input records into XML and use Xquery and Xpath ... not!

Anyway, I am interested in your thoughts on the matter. Maybe the existing RML is far as one should take the idea.

PS If I wanted to try to incorporate such extensions, where would I start?

**Comment by dachafra** [comment]
@nraphael this already exists and most of the engines implement it (RMLMapper, RocketRML). The extension incorporates the Function Ontology within RML mappings. The stable version can be seen here https://fno.io/rml/ although we are currently working on a new version https://github.com/kg-construct/rml-fno-spec 

From the scientific point of view, you can read this paper https://arxiv.org/abs/2008.13482

**Comment by nraphael** [comment]
Thank you!

Nicholas Raphael

________________________________

**Comment by dachafra** [comment]
Morph-KGC still does not implement this extension, but for the next questions regarding RML language, please consider to use https://github.com/kg-construct/rml-questions/discussions

**Comment by arenas-guerrero-julian** [comment]
Hi @nraphael ,

Thank you for starting the discussion. As @dachafra said, I think FnO functions is what you suggest. This would be a nice to have feature for Morph-KGC. However, I do not plan to include them in the short term. It would indeed require quite some work, first the [parser](https://github.com/oeg-upm/morph-kgc/blob/main/src/morph_kgc/mapping/mapping_parser.py) of the mapping would have to be extended, and then the [materialization procedure](https://github.com/oeg-upm/morph-kgc/blob/main/src/morph_kgc/materializer.py).

Julián

**Reply by nraphael** [reply]
Thank you, @ArenasGuerreroJulian . I was afraid of that. But good to know explicitly without having to study the code. Maybe one day ... the low-overhead of morph-kgc is compelling. Have you given any thought to what might be a good compromise solution, between FnO functions and straight python and/or SPARQL post-processing? Anyway, thanks again.

**Comment by arenas-guerrero-julian** [comment]
We are implementing [RML+FnO](https://kg-construct.github.io/fnml-spec/) (#117) .

**Comment by arenas-guerrero-julian** [comment]
[RML+FnO](https://kg-construct.github.io/fnml-spec/) is now supported in v2.4.0.
